src/software/thunderscope/robot_diagnostics/controller_diagnostics.py

```diff
@@ -0,0 +1,247 @@
+                "Initializing controller "
+                + self.controller.info.__str__()
+                + " and device path location: "
+                + self.controller.path
+            )
+
+            self.__stop_event_thread = Event()
+            self.__event_thread = Thread(target=self.__event_loop, daemon=True)
+            self.__event_thread.start()
+
+            self.constants = tbots_cpp.create2021RobotConstants()
+
+            self.x_vel = 0.0
+            self.y_vel = 0.0
+            self.ang_vel = 0.0
+            self.kick_power = 0.0
+            self.dribbler_speed = 0.0
+            self.dribbler_enabled = False
+
+            self.motor_control = MotorControl()
+            self.power_control = PowerControl()
+
+        else:
+            logging.debug(
+                "Tried to initialize a handheld controller from list available devices:"
+            )
+            logging.debug(list_devices())
+            logging.debug(
+                "Could not initialize a handheld controller device - check USB connections"
+            )
+
+    def get_latest_primitive_command(self):
+        if self.controller_initialized():
+            return DirectControlPrimitive(
+                motor_control=self.motor_control, power_control=self.power_control,
+            )
+        else:
+            return None
+
+    def controller_initialized(self):
+        return self.controller is not None
+
+    def __setup_controller(self):
+        for device in list_devices():
+            controller = InputDevice(device)
+            if (
+                controller is not None
+                and controller.name in ControllerConstants.VALID_CONTROLLER_NAMES
+            ):
+                self.controller = controller
+                break
+
+    @staticmethod
+    def __parse_dribbler_enabled_event_value(value) -> bool:
+        return value > (XBOX_BUTTON_MAX_RANGE / 2)
+
+    @staticmethod
+    def __parse_move_event_value(value, factor):
+        if abs(value) < (DEADZONE_PERCENTAGE * factor):
+            return 0
+        else:
+            return value / (XBOX_MAX_RANGE * factor)
+
+    def __parse_dribbler_event_value(self, value):
+        # can probably combine these 2 parsers but abastracting the +/- is annoying so 2 funcs it is
+        new_speed = self.dribbler_speed - value * DRIBBLER_STEPPER
+        if MIN_DRIBBLER_RPM < new_speed < MAX_DRIBBLER_RPM:
+            return new_speed
+        else:
+            return 0
+
+    def __parse_kick_event_value(self, value) -> int:
+        new_power = self.kick_power + value * POWER_STEPPER
+        if MIN_POWER < new_power < MAX_POWER:
+            return new_power
+        else:
+            return 0
+
+    def __process_event(self, event):
+        logging.debug("Processing event: " + str(event))
+
+        abs_event = categorize(event)
+
+        # TODO: bump python version so we can use pattern matching for this
+        if event.type == ecodes.EV_ABS:
```

Comment:
@Nima @itsarune is it hard to bump python version? It would be really nice to have access to `3.10`, they added support for the [pattern matching](https://docs.python.org/3/reference/compound_stmts.html#match) and I would be able to use it here to clean up some of the code.

Reply:
yup go ahead and bump it up `3.10` is fine

Reply:
if you want, you can bump it up to `3.12` but nima and I experienced a ton of dependency issues when we did

Reply:
Yea, I was chatting with @nimazareian about this - I will leave it as is for now, and we are thinking about starting a clean robocup branch and keeping that one prepared for comp, and master can be updated and turn into a dev branch, which it in essence already is, but I would hold off on 3.12 if it's finicky and have @Mr-Anyone put up a PR with 3.12 that we merge later and have that set-up for next year. I'll add a TODO and can make an issue that's blocked by the issue with upgrading to 3.12.